### ISSUE TITLE
Mutation issue when processing documents

### ISSUE THREAD

**Issue by achou11** [gmaclennan/realm-indexer-test#1]
I don't have write permissions to the repo and don't feel like forking this so just gonna write it all here...

Current implementation of `batch` modifies the original objects passed in as the argument:

https://github.com/gmaclennan/realm-indexer-test/blob/e8e2d389e65d9283fb13a55541444248f440665e/index.js#L52

This is causing the tests to fail because the mutations are carried over in each permutation.

A simple reproduction of this issue:

```js
const docs = [{}, {}, {}]

const indexer = {
  batch(docs) {
    for (const d of docs) {
      d.forks = []
    }
  }
}

indexer.batch(docs)

console.log(docs) // Outputs: [ { forks: [] }, { forks: [] }, { forks: [] }]
```

Suggested change (on [line 52](https://github.com/gmaclennan/realm-indexer-test/blob/e8e2d389e65d9283fb13a55541444248f440665e/index.js#L52)):

```js
const withForks = {
  ...doc, // may want to do a deep clone here
  forks: [...existing.forks.add(existing.version)],
};
realm.delete(existing);
realm.create(this.docType, withForks);
```

I tried this locally and the all tests pass after the change

**Comment by gmaclennan** [comment]
Great spot, thanks @achou11. I've got caught by this before, should probably use typescript `Readonly<>` to avoid it where possible.

**Comment by gmaclennan** [comment]
Fixed in 4dc7c3d336ccaf46b5ff0856e347fabe631a7e87
